src/main/java/microsoft/exchange/webservices/data/util/DateTimeParser.java

```diff
@@ -0,0 +1,101 @@
+ **************************************************************************/
+
+package microsoft.exchange.webservices.data.util;
+
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.TimeZone;
+
+public class DateTimeParser {
+
+  private final SimpleDateFormat[] dateTimeFormats = {
+      new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssX"),
+      new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSX"),
+      new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss"),
+      new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS"),
+      new SimpleDateFormat("yyyy-MM-ddX"),
+      new SimpleDateFormat("yyyy-MM-dd")
+  };
+
+  private final SimpleDateFormat[] dateFormats = {
+      new SimpleDateFormat("yyyy-MM-ddX"),
+      new SimpleDateFormat("yyyy-MM-dd")
+  };
+
+
+  public DateTimeParser() {
+    // Set default timezone of the formats to UTC, which will be used when the date string doesn't supply a
+    // timezone itself.
+
+    for (SimpleDateFormat format : dateTimeFormats) {
+      format.setTimeZone(TimeZone.getTimeZone("UTC"));
+    }
+
+    for (SimpleDateFormat format : dateFormats) {
+      format.setTimeZone(TimeZone.getTimeZone("UTC"));
+    }
+  }
+
+  /**
+   * Converts a date time string to local date time.
+   *
+   * Note: this method also allows dates without times, in which case the time will be 00:00:00 in the
+   * supplied timezone. UTC timezone will be assumed if no timezone is supplied.
+   *
+   * @param value The string value to parse.
+   * @return The parsed {@link Date}.
+   *
+   * @throws java.lang.IllegalArgumentException If string can not be parsed.
+   */
+  public Date convertDateTimeStringToDate(String value) {
+    return parseInternal(value, false);
+  }
+
+  /**
+   * Converts a date string to local date time.
+   *
+   * UTC timezone will be assumed if no timezone is supplied.
+   *
+   * @param value The string value to parse.
+   * @return The parsed {@link Date}.
+   *
+   * @throws java.lang.IllegalArgumentException If string can not be parsed.
+   */
+  public Date convertDateStringToDate(String value) {
+    return parseInternal(value, true);
+  }
+
+  private Date parseInternal(String value, boolean dateOnly) {
+    String originalValue = value;
+
+    if (value == null || value.isEmpty()) {
+      return null;
+    } else {
+      if (value.endsWith("z")) {
+        // This seems to be an edge case. Let's uppercase the Z to be sure.
+        value = value.substring(0, value.length() - 1) + "Z";
+      }
+
+      SimpleDateFormat[] formats = dateOnly ? dateFormats : dateTimeFormats;
+      for (SimpleDateFormat format : formats) {
+        try {
+          return format.parse(value);
+        } catch (ParseException e) {
+          // Ignore and try the next pattern.
+        }
+      }
+    }
+
+    throw new IllegalArgumentException(
```

Comment:
Yep, I've just read about [checked vs. runtime exceptions](http://johnpwood.net/2008/04/21/java-checked-exceptions-vs-runtime-exceptions/), I'm new to the concept of checked exceptions since it is unique to Java.  However, I would have thought that parsing should be a checked exception to force the callers to handle the case where the string they are providing is not valid.  This is not something that you want to bubble to the top of the stack.  It is interesting however that SimpleDateFormat throws ParseException (checked) and long.ParseLong throws IllegalArgumentException (runtime).  I would have thought there would be consistency there and both use checked, since I would expect parsing errors to be handled closest to where they are used.